sql/importer: TestImportWorkerFailure failed

sql/importer.TestImportWorkerFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9959373?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9959373?buildTab=artifacts#/) on release-23.1 @ [4226a83871bbce776bc9389fca5cf084b4bb7632](https://github.com/cockroachdb/cockroach/commits/4226a83871bbce776bc9389fca5cf084b4bb7632):


```
=== RUN   TestImportWorkerFailure
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/73dfce30b9a5630b1b4dabed3c94b32c/logTestImportWorkerFailure3379689604
    test_log_scope.go:79: use -show-logs to present logs inline
    import_stmt_test.go:5430: pq: updating job details after publishing schemas: job 862797201834016769: select-job: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/73dfce30b9a5630b1b4dabed3c94b32c/logTestImportWorkerFailure3379689604
--- FAIL: TestImportWorkerFailure (42.46s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestImportWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27679

Using stress after 2885 runs I got:

```
test logs left over in: /tmp/_tmp/060246e25d27a0efc6dd14ec672a0f2b/logTestImportWorkerFailure3155831314
--- FAIL: TestImportWorkerFailure (17.54s)
    test_log_scope.go:161: test logs captured to: /tmp/_tmp/060246e25d27a0efc6dd14ec672a0f2b/logTestImportWorkerFailure3155831314
    test_log_scope.go:79: use -show-logs to present logs inline
    import_stmt_test.go:5435: query 'SELECT * FROM t ORDER BY i': expected:
        0
        1
        2
        3
        4
        5
        6
        7
        8
        9
        10
        11
        12
        13
        14
        15
        16
        17
        18
        19
        
        got:
        1
        2
        4
        5
        7
        8
        10
        11
        13
        14
        16
        17
        19
        
    panic.go:522: -- test log scope end --
FAIL
I230509 15:29:22.694464 1 (gostd) testmain.go:254  [-] 1  Test //pkg/sql/importer:importer_test exited with error code 1
```

sql/importer.TestImportWorkerFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10361498?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10361498?buildTab=artifacts#/) on release-23.1 @ [e84e2cabef32a59d88041132f059a6768dfe56bb](https://github.com/cockroachdb/cockroach/commits/e84e2cabef32a59d88041132f059a6768dfe56bb):


```
=== RUN   TestImportWorkerFailure
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/73dfce30b9a5630b1b4dabed3c94b32c/logTestImportWorkerFailure1950566143
    test_log_scope.go:79: use -show-logs to present logs inline
    import_stmt_test.go:5443: pq: updating job details after publishing schemas: job 870445175802920961: select-job: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
    panic.go:522: -- test log scope end --
--- FAIL: TestImportWorkerFailure (43.97s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestImportWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/importer.TestImportWorkerFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10937598?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10937598?buildTab=artifacts#/) on release-23.1 @ [78dae31f503cec8e00fa2f18ed6a65da6042acbe](https://github.com/cockroachdb/cockroach/commits/78dae31f503cec8e00fa2f18ed6a65da6042acbe):


```
=== RUN   TestImportWorkerFailure
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/73dfce30b9a5630b1b4dabed3c94b32c/logTestImportWorkerFailure2654984900
    test_log_scope.go:79: use -show-logs to present logs inline
    import_stmt_test.go:5439: pq: updating job details after publishing schemas: job 883166950884081665: select-job: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/73dfce30b9a5630b1b4dabed3c94b32c/logTestImportWorkerFailure2654984900
--- FAIL: TestImportWorkerFailure (67.07s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestImportWorkerFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/importer/import_stmt_test.go
+++ b/pkg/sql/importer/import_stmt_test.go
@@ -5417,9 +5417,9 @@ func TestImportWorkerFailure(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.UnderStressWithIssue(t, 102839, "flaky test")
-	skip.UnderDeadlockWithIssue(t, 102839, "flaky test")
-	skip.UnderRaceWithIssue(t, 102839, "flaky test")
+	skip.UnderStressWithIssue(t, 108547, "flaky test")
+	skip.UnderDeadlockWithIssue(t, 108547, "flaky test")
+	skip.UnderRaceWithIssue(t, 108547, "flaky test")
 
 	allowResponse := make(chan struct{})
 	params := base.TestClusterArgs{}
@@ -5468,10 +5468,15 @@ func TestImportWorkerFailure(t *testing.T) {
 	tc.StopServer(1)
 
 	close(allowResponse)
-	// We expect the statement to retry since it should have encountered a
-	// retryable error.
+	// We expect the IMPORT statement to usually retry since it should have
+	// encountered a retryable error. We don't currently catch all such retryable
+	// errors, however, so in some cases the IMPORT statement will return the
+	// error to the client. In this case we verify that the import was completely
+	// rolled back.
 	if err := <-errCh; err != nil {
-		t.Fatal(err)
+		t.Logf("%s failed, checking that imported data was completely removed: %q", query, err)
+		sqlDB.CheckQueryResults(t, `SELECT * FROM t ORDER BY i`, [][]string{})
+		return
 	}
 
 	// But the job should be restarted and succeed eventually.

